osu.Game.Rulesets.Osu/Edit/OsuHitObjectComposer.cs

```diff
@@ -32,9 +37,66 @@ protected override DrawableRuleset<OsuHitObject> CreateDrawableRuleset(Ruleset r
             new SpinnerCompositionTool()
         };
 
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            EditorBeatmap.SelectedHitObjects.CollectionChanged += (_, __) => updateDistanceSnapGrid();
+            EditorBeatmap.PlacementObject.ValueChanged += _ => updateDistanceSnapGrid();
+
+            LayerBelowRuleset.Add(distanceSnapGridContainer = new Container { RelativeSizeAxes = Axes.Both });
+        }
+
         protected override ComposeBlueprintContainer CreateBlueprintContainer() => new OsuBlueprintContainer(HitObjects);
 
-        protected override DistanceSnapGrid CreateDistanceSnapGrid(IEnumerable<HitObject> selectedHitObjects)
+        private DistanceSnapGrid distanceSnapGrid;
+        private Container distanceSnapGridContainer;
+
+        private readonly Cached distanceSnapGridCache = new Cached();
+        private double? lastDistanceSnapGridTime;
+
+        protected override void Update()
+        {
+            base.Update();
+
+            if (!(BlueprintContainer.CurrentTool is SelectTool))
+            {
+                if (EditorClock.CurrentTime != lastDistanceSnapGridTime)
+                {
+                    distanceSnapGridCache.Invalidate();
+                    lastDistanceSnapGridTime = EditorClock.CurrentTime;
+                }
+
+                if (!distanceSnapGridCache.IsValid)
+                    updateDistanceSnapGrid();
+            }
+        }
+
+        public override SnapResult SnapScreenSpacePositionToValidTime(Vector2 screenSpacePosition)
+        {
+            if (distanceSnapGrid == null)
+                return base.SnapScreenSpacePositionToValidTime(screenSpacePosition);
+
+            (Vector2 pos, double time) = distanceSnapGrid.GetSnappedPosition(distanceSnapGrid.ToLocalSpace(screenSpacePosition));
+
+            return new SnapResult(distanceSnapGrid.ToScreenSpace(pos), time);
+        }
+
+        private void updateDistanceSnapGrid()
```

Comment:
This also needs to be called when the select tool is selected. Without that the placement grid remains visible.

Reply:
Aha. I was only testing using 1 / 2 hotkeys, where it works as expected. Not 100% happy on the invalidation logic for this

Reply:
Fix in a better feeling way – now the grid will disappear when exiting the placement area.